### ISSUE TITLE
fenics package not installable

### ISSUE THREAD

**Issue by grisuthedragon** [spack/spack#7145]
If you want to report a build error for some package, or a bug in core please use the following template as a guideline:

-----

Summary

### Expected Result

Get FEniCS, a nice python/C++ finite element package working with spack. The package is already included in the builtin repository. 

### Actual Result

In a clean enviroment, without any install spack-package it the build fails with: 
```
$ spack install -v fenics@2016.1.0
==> Error: An unsatisfiable variant constraint has been detected for spec:

    hdf5@1.10.1%gcc@5.4.0~cxx~debug~fortran~hl+mpi+pic+shared~szip~threadsafe arch=linux-ubuntu16.04-x86_64 
        ^openmpi@3.0.0%gcc@5.4.0~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple~ucx+vt arch=linux-ubuntu16.04-x86_64 
            ^hwloc
                ^pkgconf@1.4.0%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
        ^zlib@1.2.11%gcc@5.4.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64 


while trying to concretize the partial spec:

    netcdf+mpi
        ^m4@1.4.18%gcc@5.4.0+sigsegv arch=linux-ubuntu16.04-x86_64 
            ^libsigsegv
        ^openmpi@3.0.0%gcc@5.4.0~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple~ucx+vt arch=linux-ubuntu16.04-x86_64 
            ^hwloc
                ^pkgconf@1.4.0%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
        ^zlib@1.2.11%gcc@5.4.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64 


netcdf requires hdf5 variant +hl, but spec asked for ~hl
```

If the hdf5 support for fenics is turned of the error moves on to the petsc dependency: 
```
$ spack install -v fenics@2016.1.0 -hdf5
==> Error: An unsatisfiable variant constraint has been detected for spec:

    metis@5.1.0%gcc@5.4.0 build_type=Release ~gdb~int64~real64+shared arch=linux-ubuntu16.04-x86_64 
        ^cmake@3.10.1%gcc@5.4.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu16.04-x86_64 
            ^ncurses@6.0%gcc@5.4.0~symlinks~termlib arch=linux-ubuntu16.04-x86_64 
                ^pkgconf@1.4.0%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
            ^openssl@1.0.2n%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
                ^zlib@1.2.11%gcc@5.4.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64 


while trying to concretize the partial spec:

    petsc@3.8.3%gcc@5.4.0+boost clanguage=C ~complex~debug+double+hdf5+hypre~int64+metis+mpi~mumps+shared+superlu-dist~trilinos arch=linux-ubuntu16.04-x86_64 
        ^hdf5@1.8.9:+hl+mpi
            ^zlib@1.2.11%gcc@5.4.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64 
        ^hypre~int64~internal-superlu
            ^openblas@0.2.20%gcc@5.4.0 cpu_target= ~ilp64+pic+shared threads=none ~virtual_machine arch=linux-ubuntu16.04-x86_64 
        ^superlu-dist@5.2:5.2.99~int64
            ^metis@5.1.0%gcc@5.4.0 build_type=Release ~gdb~int64~real64+shared arch=linux-ubuntu16.04-x86_64 
                ^cmake@3.10.1%gcc@5.4.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu16.04-x86_64 
                    ^ncurses@6.0%gcc@5.4.0~symlinks~termlib arch=linux-ubuntu16.04-x86_64 
                        ^pkgconf@1.4.0%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
                    ^openssl@1.0.2n%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
            ^openmpi@3.0.0%gcc@5.4.0~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple~ucx+vt arch=linux-ubuntu16.04-x86_64 
                ^hwloc
            ^parmetis@4.0.3%gcc@5.4.0 build_type=RelWithDebInfo ~gdb+shared arch=linux-ubuntu16.04-x86_64 


```

### Steps to reproduce the issue
Create an empty spack installation from the current git repository.
```console
$ spack install fenics@2016.1.0
$ spack install fenics@2016.1.0 -hdf5
```

### Information on your system
I work on a 

- Ubuntu 16.04 with gcc 5.4.0 (detected by spack) 
- x86_64
- no customizations in spack or its config.

**Comment by davydden** [comment]
try
```
diff --git a/var/spack/repos/builtin/packages/fenics/package.py b/var/spack/repos/builtin/packages/fenics/package.py
index d95c826c3..4c6522ac3 100644
--- a/var/spack/repos/builtin/packages/fenics/package.py
+++ b/var/spack/repos/builtin/packages/fenics/package.py
@@ -77,7 +77,7 @@ class Fenics(CMakePackage):
     depends_on('boost+filesystem+program_options+system+iostreams+timer+regex+chrono')
 
     depends_on('mpi', when='+mpi')
-    depends_on('hdf5', when='+hdf5')
+    depends_on('hdf5+hl', when='+hdf5')
     depends_on('parmetis@4.0.2:^metis+real64', when='+parmetis')
     depends_on('scotch~metis', when='+scotch~mpi')
     depends_on('scotch+mpi~metis', when='+scotch+mpi')
```
if it helps, feel free to create a PR with the fix.

p.s. this a known bug in concretizer

**Comment by grisuthedragon** [comment]
The patch only helps against the hdf5 problem and netcdf. If I apply the patch it get in struggle with the petsc package as shown above. Is this a similar kind of bug?

**Comment by davydden** [comment]
@grisuthedragon I don't see any issues with this patch:

from 6170ca863dd8f3cb7c1e0e86f51fcb6a097ce421 (Jan20 2018)

```
$ spack spec fenics
Input spec
--------------------------------
fenics

Concretized
--------------------------------
fenics@2016.1.0%clang@9.0.0-apple build_type=RelWithDebInfo ~doc+hdf5+mpi+openmp+parmetis patches=8de9625a4ba4d27801bde734f744c96efcce57ecdec63556a5f6e2c28d4af532 +petsc~qt+scotch+shared+slepc+suite-sparse+trilinos~vtk arch=darwin-highsierra-x86_64
    ^boost@1.66.0%clang@9.0.0-apple+atomic+chrono~clanglibcpp+date_time~debug+exception+filesystem~graph~icu+iostreams+locale+log+math~mpi+multithreaded+program_options+python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=darwin-highsierra-x86_64
        ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-highsierra-x86_64
        ^python@2.7.14%clang@9.0.0-apple patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic~shared~tk~ucs4 arch=darwin-highsierra-x86_64
            ^ncurses@6.0%clang@9.0.0-apple patches=f84b2708a42777aadcc7f502a261afe10ca5646a51c1ef8b5e60d2070d926b57 ~symlinks~termlib arch=darwin-highsierra-x86_64
                ^pkgconf@1.4.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^openssl@1.0.2k%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64
            ^readline@7.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^sqlite@3.20.0%clang@9.0.0-apple patches=5bbcba091045e547eb550e6e9b9372009c41be0a9de8a5c030d03361c11ff939 arch=darwin-highsierra-x86_64
    ^cmake@3.10.1%clang@9.0.0-apple~doc+ncurses+openssl+ownlibs~qt arch=darwin-highsierra-x86_64
    ^eigen@3.3.3%clang@9.0.0-apple build_type=RelWithDebInfo +fftw+metis+mpfr+scotch+suitesparse arch=darwin-highsierra-x86_64
        ^fftw@3.3.7%clang@9.0.0-apple+double+float+long_double+mpi~openmp~pfft_patches~quad arch=darwin-highsierra-x86_64
            ^openmpi@3.0.0%clang@9.0.0-apple~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple~ucx+vt arch=darwin-highsierra-x86_64
                ^hwloc@1.11.8%clang@9.0.0-apple~cuda+libxml2~pci+shared arch=darwin-highsierra-x86_64
                    ^libxml2@2.9.4%clang@9.0.0-apple~python arch=darwin-highsierra-x86_64
                        ^xz@5.2.3%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^gmp@6.1.2%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^autoconf@2.69%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^m4@1.4.18%clang@9.0.0-apple patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00 +sigsegv arch=darwin-highsierra-x86_64
                    ^libsigsegv@2.11%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^perl@5.24.1%clang@9.0.0-apple+cpanm+shared arch=darwin-highsierra-x86_64
                    ^gdbm@1.14.1%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^automake@1.15.1%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^libtool@2.4.6%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^metis@5.1.0%clang@9.0.0-apple build_type=Release ~gdb~int64 patches=4991da938c1d3a1d3dea78e49bbebecba00273f98df2a656e38b83d55b281da1 +real64+shared arch=darwin-highsierra-x86_64
        ^mpfr@3.1.5%clang@9.0.0-apple patches=88dfefa6d39c9fd5a26a40d9bbc73df8da93f946c65c06038026261d78d919f5,dfd62a42a239c427bb44087b4a520edbcd65bae6ad1fbe07303c96ee8696afbd arch=darwin-highsierra-x86_64
        ^scotch@6.0.4%clang@9.0.0-apple+compression~esmumps~int64~metis+mpi patches=740b3e01d619b030a415f96548a762a09a1af06d498cef7bec8579a4fb11abd8,fe20c09b006f8e59f31a16e85cc43ac786eecf7de0348e5c9c57685c2afbc86e +shared arch=darwin-highsierra-x86_64
            ^bison@3.0.4%clang@9.0.0-apple patches=57f972940a10d448efbd3d5ba46e65979ae4eea93681a85e1d998060b356e0d2,b72914fe38e54a6fc25f29019e0a0786705c4f61ce20d414cc2010c8d62448c7 arch=darwin-highsierra-x86_64
            ^flex@2.6.4%clang@9.0.0-apple+lex arch=darwin-highsierra-x86_64
                ^gettext@0.19.8.1%clang@9.0.0-apple+bzip2+curses+git~libunistring+libxml2 patches=9acdb4e73f67c241b5ef32505c9ddf7cf6884ca8ea661692f21dca28483b04b8 +tar+xz arch=darwin-highsierra-x86_64
                    ^tar@1.29%clang@9.0.0-apple patches=08921fcbd732050c74ddf1de7d8ad95ffdbc09f8b4342456fa2f6a0dd02a957c,5c314db58d005043bf407abaf25eb9823b9032a22fd12a0b142d4bf548130fa4,d428578be7fb99b831eb61e53b8d88a859afe08b479a21238180899707d79ce4,e34d2470461037933737c0a60fbcc5d5ea905fefb6b414b4a2e88a440fcae23a arch=darwin-highsierra-x86_64
                ^help2man@1.47.4%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^suite-sparse@5.1.0%clang@9.0.0-apple~cuda~openmp+pic~tbb arch=darwin-highsierra-x86_64
            ^openblas@0.2.20%clang@9.0.0-apple cpu_target= ~ilp64 patches=47cfa7a952ac7b2e4632c73ae199d69fb54490627b66a62c681e21019c4ddc9d +pic+shared threads=none ~virtual_machine arch=darwin-highsierra-x86_64
    ^hdf5@1.10.1%clang@9.0.0-apple~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=darwin-highsierra-x86_64
    ^parmetis@4.0.3%clang@9.0.0-apple build_type=RelWithDebInfo ~gdb patches=4f892531eb0a807eb1b82e683a416d3e35154a455274cf9b162fb02054d11a5b,50ed2081bc939269689789942067c58b3e522c269269a430d5d34c00edbc5870,704b84f7c7444d4372cb59cca6e1209df4ef3b033bc4ee3cf50f369bce972a9d +shared arch=darwin-highsierra-x86_64
    ^petsc@3.8.3%clang@9.0.0-apple+boost clanguage=C ~complex~debug+double+hdf5+hypre~int64+metis+mpi~mumps+shared+superlu-dist~trilinos arch=darwin-highsierra-x86_64
        ^hypre@2.13.0%clang@9.0.0-apple~int64~internal-superlu+mpi~shared arch=darwin-highsierra-x86_64
        ^superlu-dist@5.2.2%clang@9.0.0-apple~int64 arch=darwin-highsierra-x86_64
    ^py-numpy@1.13.3%clang@9.0.0-apple+blas+lapack arch=darwin-highsierra-x86_64
        ^py-setuptools@35.0.2%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^py-appdirs@1.4.3%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^py-packaging@16.8%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^py-pyparsing@2.2.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^py-six@1.10.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^py-ply@3.8%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^py-sympy@1.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^py-mpmath@0.19%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^slepc@3.8.2%clang@9.0.0-apple+arpack arch=darwin-highsierra-x86_64
        ^arpack-ng@3.5.0%clang@9.0.0-apple+mpi+shared arch=darwin-highsierra-x86_64
    ^swig@3.0.12%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^pcre@8.41%clang@9.0.0-apple~jit+utf arch=darwin-highsierra-x86_64
    ^trilinos@12.12.1%clang@9.0.0-apple~alloptpkgs+amesos+amesos2+anasazi+aztec+belos+boost build_type=RelWithDebInfo ~cgns~dtk+epetra+epetraext+exodus+fortran~fortrilinos+gtest+hdf5+hypre+ifpack+ifpack2+instantiate~instantiate_cmplx~intrepid~intrepid2+metis+ml+muelu+mumps~nox~openmp~pnetcdf~python+rol+sacado~shards+shared~stk+suite-sparse~superlu+superlu-dist+teuchos+tpetra~x11~xsdkflags~zlib+zoltan+zoltan2 arch=darwin-highsierra-x86_64
        ^glm@0.9.7.1%clang@9.0.0-apple build_type=RelWithDebInfo arch=darwin-highsierra-x86_64
        ^matio@1.5.9%clang@9.0.0-apple+hdf5+shared+zlib arch=darwin-highsierra-x86_64
        ^mumps@5.1.1%clang@9.0.0-apple+complex+double+float~int64~metis+mpi~parmetis~ptscotch~scotch+shared arch=darwin-highsierra-x86_64
            ^netlib-scalapack@2.0.2%clang@9.0.0-apple build_type=RelWithDebInfo ~pic+shared arch=darwin-highsierra-x86_64
        ^netcdf@4.4.1.1%clang@9.0.0-apple~dap~hdf4 maxdims=1024 maxvars=8192 +mpi~parallel-netcdf+shared arch=darwin-highsierra-x86_64

faustaff-010-020-010-084:spack davydden$ spack spec fenics
Input spec
--------------------------------
fenics

Concretized
--------------------------------
fenics@2016.1.0%clang@9.0.0-apple build_type=RelWithDebInfo ~doc+hdf5+mpi+openmp+parmetis patches=8de9625a4ba4d27801bde734f744c96efcce57ecdec63556a5f6e2c28d4af532 +petsc~qt+scotch+shared+slepc+suite-sparse+trilinos~vtk arch=darwin-highsierra-x86_64
    ^boost@1.66.0%clang@9.0.0-apple+atomic+chrono~clanglibcpp+date_time~debug+exception+filesystem~graph~icu+iostreams+locale+log+math~mpi+multithreaded+program_options+python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=darwin-highsierra-x86_64
        ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-highsierra-x86_64
        ^python@2.7.14%clang@9.0.0-apple patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic~shared~tk~ucs4 arch=darwin-highsierra-x86_64
            ^ncurses@6.0%clang@9.0.0-apple patches=f84b2708a42777aadcc7f502a261afe10ca5646a51c1ef8b5e60d2070d926b57 ~symlinks~termlib arch=darwin-highsierra-x86_64
                ^pkgconf@1.4.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^openssl@1.0.2k%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64
            ^readline@7.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^sqlite@3.20.0%clang@9.0.0-apple patches=5bbcba091045e547eb550e6e9b9372009c41be0a9de8a5c030d03361c11ff939 arch=darwin-highsierra-x86_64
    ^cmake@3.10.1%clang@9.0.0-apple~doc+ncurses+openssl+ownlibs~qt arch=darwin-highsierra-x86_64
    ^eigen@3.3.3%clang@9.0.0-apple build_type=RelWithDebInfo +fftw+metis+mpfr+scotch+suitesparse arch=darwin-highsierra-x86_64
        ^fftw@3.3.7%clang@9.0.0-apple+double+float+long_double+mpi~openmp~pfft_patches~quad arch=darwin-highsierra-x86_64
            ^openmpi@3.0.0%clang@9.0.0-apple~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple~ucx+vt arch=darwin-highsierra-x86_64
                ^hwloc@1.11.8%clang@9.0.0-apple~cuda+libxml2~pci+shared arch=darwin-highsierra-x86_64
                    ^libxml2@2.9.4%clang@9.0.0-apple~python arch=darwin-highsierra-x86_64
                        ^xz@5.2.3%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^gmp@6.1.2%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^autoconf@2.69%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^m4@1.4.18%clang@9.0.0-apple patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00 +sigsegv arch=darwin-highsierra-x86_64
                    ^libsigsegv@2.11%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^perl@5.24.1%clang@9.0.0-apple+cpanm+shared arch=darwin-highsierra-x86_64
                    ^gdbm@1.14.1%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^automake@1.15.1%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^libtool@2.4.6%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^metis@5.1.0%clang@9.0.0-apple build_type=Release ~gdb~int64 patches=4991da938c1d3a1d3dea78e49bbebecba00273f98df2a656e38b83d55b281da1 +real64+shared arch=darwin-highsierra-x86_64
        ^mpfr@3.1.5%clang@9.0.0-apple patches=88dfefa6d39c9fd5a26a40d9bbc73df8da93f946c65c06038026261d78d919f5,dfd62a42a239c427bb44087b4a520edbcd65bae6ad1fbe07303c96ee8696afbd arch=darwin-highsierra-x86_64
        ^scotch@6.0.4%clang@9.0.0-apple+compression~esmumps~int64~metis+mpi patches=740b3e01d619b030a415f96548a762a09a1af06d498cef7bec8579a4fb11abd8,fe20c09b006f8e59f31a16e85cc43ac786eecf7de0348e5c9c57685c2afbc86e +shared arch=darwin-highsierra-x86_64
            ^bison@3.0.4%clang@9.0.0-apple patches=57f972940a10d448efbd3d5ba46e65979ae4eea93681a85e1d998060b356e0d2,b72914fe38e54a6fc25f29019e0a0786705c4f61ce20d414cc2010c8d62448c7 arch=darwin-highsierra-x86_64
            ^flex@2.6.4%clang@9.0.0-apple+lex arch=darwin-highsierra-x86_64
                ^gettext@0.19.8.1%clang@9.0.0-apple+bzip2+curses+git~libunistring+libxml2 patches=9acdb4e73f67c241b5ef32505c9ddf7cf6884ca8ea661692f21dca28483b04b8 +tar+xz arch=darwin-highsierra-x86_64
                    ^tar@1.29%clang@9.0.0-apple patches=08921fcbd732050c74ddf1de7d8ad95ffdbc09f8b4342456fa2f6a0dd02a957c,5c314db58d005043bf407abaf25eb9823b9032a22fd12a0b142d4bf548130fa4,d428578be7fb99b831eb61e53b8d88a859afe08b479a21238180899707d79ce4,e34d2470461037933737c0a60fbcc5d5ea905fefb6b414b4a2e88a440fcae23a arch=darwin-highsierra-x86_64
                ^help2man@1.47.4%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^suite-sparse@5.1.0%clang@9.0.0-apple~cuda~openmp+pic~tbb arch=darwin-highsierra-x86_64
            ^openblas@0.2.20%clang@9.0.0-apple cpu_target= ~ilp64 patches=47cfa7a952ac7b2e4632c73ae199d69fb54490627b66a62c681e21019c4ddc9d +pic+shared threads=none ~virtual_machine arch=darwin-highsierra-x86_64
    ^hdf5@1.10.1%clang@9.0.0-apple~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=darwin-highsierra-x86_64
    ^parmetis@4.0.3%clang@9.0.0-apple build_type=RelWithDebInfo ~gdb patches=4f892531eb0a807eb1b82e683a416d3e35154a455274cf9b162fb02054d11a5b,50ed2081bc939269689789942067c58b3e522c269269a430d5d34c00edbc5870,704b84f7c7444d4372cb59cca6e1209df4ef3b033bc4ee3cf50f369bce972a9d +shared arch=darwin-highsierra-x86_64
    ^petsc@3.8.3%clang@9.0.0-apple+boost clanguage=C ~complex~debug+double+hdf5+hypre~int64+metis+mpi+mumps+shared+superlu-dist~trilinos arch=darwin-highsierra-x86_64
        ^hypre@2.13.0%clang@9.0.0-apple~int64~internal-superlu+mpi~shared arch=darwin-highsierra-x86_64
        ^mumps@5.1.1%clang@9.0.0-apple+complex+double+float~int64~metis+mpi~parmetis~ptscotch~scotch+shared arch=darwin-highsierra-x86_64
            ^netlib-scalapack@2.0.2%clang@9.0.0-apple build_type=RelWithDebInfo ~pic+shared arch=darwin-highsierra-x86_64
        ^superlu-dist@5.2.2%clang@9.0.0-apple~int64 arch=darwin-highsierra-x86_64
    ^py-numpy@1.13.3%clang@9.0.0-apple+blas+lapack arch=darwin-highsierra-x86_64
        ^py-setuptools@35.0.2%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^py-appdirs@1.4.3%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^py-packaging@16.8%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^py-pyparsing@2.2.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^py-six@1.10.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^py-ply@3.8%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^py-sympy@1.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^py-mpmath@0.19%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^slepc@3.8.2%clang@9.0.0-apple+arpack arch=darwin-highsierra-x86_64
        ^arpack-ng@3.5.0%clang@9.0.0-apple+mpi+shared arch=darwin-highsierra-x86_64
    ^swig@3.0.12%clang@9.0.0-apple arch=darwin-highsierra-x86_64
        ^pcre@8.41%clang@9.0.0-apple~jit+utf arch=darwin-highsierra-x86_64
    ^trilinos@12.12.1%clang@9.0.0-apple~alloptpkgs+amesos+amesos2+anasazi+aztec+belos+boost build_type=RelWithDebInfo ~cgns~dtk+epetra+epetraext+exodus+fortran~fortrilinos+gtest+hdf5+hypre+ifpack+ifpack2+instantiate~instantiate_cmplx~intrepid~intrepid2+metis+ml+muelu+mumps~nox~openmp~pnetcdf~python+rol+sacado~shards+shared~stk+suite-sparse~superlu+superlu-dist+teuchos+tpetra~x11~xsdkflags~zlib+zoltan+zoltan2 arch=darwin-highsierra-x86_64
        ^glm@0.9.7.1%clang@9.0.0-apple build_type=RelWithDebInfo arch=darwin-highsierra-x86_64
        ^matio@1.5.9%clang@9.0.0-apple+hdf5+shared+zlib arch=darwin-highsierra-x86_64
        ^netcdf@4.4.1.1%clang@9.0.0-apple~dap~hdf4 maxdims=1024 maxvars=8192 +mpi~parallel-netcdf+shared arch=darwin-highsierra-x86_64
```

**Comment by grisuthedragon** [comment]
After applying the patch I get 
```
$ spack spec fenics
Input spec
--------------------------------
fenics

Concretized
--------------------------------
==> Error: An unsatisfiable variant constraint has been detected for spec:

    metis@5.1.0%gcc@5.4.0 build_type=Release ~gdb~int64~real64+shared arch=linux-ubuntu16.04-x86_64 
        ^cmake@3.10.1%gcc@5.4.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu16.04-x86_64 
            ^ncurses@6.0%gcc@5.4.0~symlinks~termlib arch=linux-ubuntu16.04-x86_64 
                ^pkgconf@1.4.0%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
            ^openssl@1.0.2n%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
                ^zlib@1.2.11%gcc@5.4.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64 


while trying to concretize the partial spec:

    petsc@3.8.3%gcc@5.4.0+boost clanguage=C ~complex~debug+double+hdf5+hypre~int64+metis+mpi~mumps+shared+superlu-dist~trilinos arch=linux-ubuntu16.04-x86_64 
        ^hdf5@1.10.1%gcc@5.4.0~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=linux-ubuntu16.04-x86_64 
            ^openmpi@3.0.0%gcc@5.4.0~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple~ucx+vt arch=linux-ubuntu16.04-x86_64 
                ^hwloc
                    ^pkgconf@1.4.0%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
            ^zlib@1.2.11%gcc@5.4.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64 
        ^hypre~int64~internal-superlu
            ^openblas@0.2.20%gcc@5.4.0 cpu_target= ~ilp64+pic+shared threads=none ~virtual_machine arch=linux-ubuntu16.04-x86_64 
        ^superlu-dist@5.2:5.2.99~int64
            ^metis@5.1.0%gcc@5.4.0 build_type=Release ~gdb~int64~real64+shared arch=linux-ubuntu16.04-x86_64 
                ^cmake@3.10.1%gcc@5.4.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu16.04-x86_64 
                    ^ncurses@6.0%gcc@5.4.0~symlinks~termlib arch=linux-ubuntu16.04-x86_64 
                    ^openssl@1.0.2n%gcc@5.4.0 arch=linux-ubuntu16.04-x86_64 
            ^parmetis@4.0.3%gcc@5.4.0 build_type=RelWithDebInfo ~gdb+shared arch=linux-ubuntu16.04-x86_64 


petsc requires metis variant +real64, but spec asked for ~real64

```

**Comment by davydden** [comment]
@grisuthedragon are you sure you are running the latest version of Spack? That's the reason I explicitly mentioned the hash on which I tested this patch.

**Comment by grisuthedragon** [comment]
I am on  cb5919b which is the current head of the develop branch.

**Comment by davydden** [comment]
I see why it worked for me, in `packages.yaml` i have
```
  metis:
    variants: +real64
```

so the second part of the changes should be along the lines of https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dealii/package.py#L131-L134

In other words, one has to be explicit about `metis` variant. 

@tgamblin just wanted to remind you about this bug in concretizer, IMO it has been breaking quite a number of packages.

**Comment by grisuthedragon** [comment]
This separate entry seems to help. Lets see if everything compiles fine.

**Comment by davydden** [comment]
> This separate entry seems to help.

ideally, one should adjust `fenics` package so that this bug does not appear. Otherwise everyone else who tries it will see the same problem.

**Comment by davydden** [comment]
>  But the other one, regarding metis, seems to be somewhere deep inside the petsc package as you showed for deal.II as well.

no no, it's not in `petsc` package. It has to do with how concretization is being done. So the right approach for now  is to modify `fenics` packages similar to `dealii`, i.e. be explicit about `+real64` variant in `metis` dependency.

**Comment by mathstuf** [comment]
I think the core issue is #593.

**Comment by alalazo** [comment]
Should be solved by #7404, please reopen if the issue persists.
